### PR TITLE
Remove TypedMorphirRuntimeDefs

### PR DESCRIPTION
**Pull request by Bouzomgi**
Removes all of TypedMorphirRuntimeDefs.scala, which housed alias types that duplicated what we already have as TypedValue and TypedDefinition

### CODE REVIEW

File: morphir/runtime/src/org/finos/morphir/runtime/MorphirRuntime.scala [hunk 1]
```diff
@@ -11,32 +11,29 @@ import org.finos.morphir.runtime.quick.QuickMorphirRuntime
...
 
 trait MorphirRuntime {
-  type TypeAttribs
-  type ValueAttribs
-
   def evaluate(
-      entryPoint: Value[TypeAttribs, ValueAttribs],
-      param: Value[TypeAttribs, ValueAttribs],
-      params: Value[TypeAttribs, ValueAttribs]*
+      entryPoint: Value[scala.Unit, UType],
```
Comment (DamianReeves) [inline morphir/runtime/src/org/finos/morphir/runtime/MorphirRuntime.scala:15]: Why are we doing this? This goes against the concept of MorphirRuntime

**Comment by DamianReeves** [comment]
I'm actually gonna question the very existence of this PR. Someone is not understanding something fundamental here.

File: morphir/src/org/finos/morphir/runtime/TypedMorphirRuntimeDefs.scala [hunk 2]
```diff
@@ -1,11 +0,0 @@
...
-
-import org.finos.morphir.ir.Type.UType
-import org.finos.morphir.ir.Value.Value
-
-object TypedMorphirRuntimeDefs {
-  final type TypeAttribs  = scala.Unit
-  final type ValueAttribs = UType
-  type RuntimeValue       = Value[TypeAttribs, ValueAttribs]
-  type RuntimeDefinition  = Value.Definition[TypeAttribs, ValueAttribs]
-}
```
Comment (edwardpeters) [inline morphir/src/org/finos/morphir/runtime/TypedMorphirRuntimeDefs.scala:11]: @DamianReeves 

There's definitely been some miscommunication here. This change - the deletion of this file - was what I expected to see. Outside of that, my understanding was that everything that referenced types this file should be refactored to point to type definitions from `morphir/src/org/finos/morphir/ir/Value.scala` or `morphir/src/org/finos/morphir/ir/Type.scala` instead. Is that in line with your expectations?

One issue is anything that referenced `TypedMorphirRuntimeDefs.TypeAttribs` - that I did expect to be replaced with `scala.Unit`, given that there is no equivalent alias. However, complicating matters is that it looks like some places defined things as `Value[TypedMorphirRuntimeDefs.TypeAttribs, TypedMorphirRuntimeDefs.ValueAttribs]` , which was probably a pre-existing issue - those probably should have defined things as `TypedMorphirRuntimeDefs.RuntimeValue`, which could be replaced by `Value.TypedValue`. I'm not 100% sure, but I think that would eliminate all hard-coded uses of `scala.Unit` as a type in the runtime.

File: morphir/runtime/src/org/finos/morphir/runtime/MorphirRuntime.scala (hunk 1 above)
Comment (edwardpeters) [reply morphir/runtime/src/org/finos/morphir/runtime/MorphirRuntime.scala:15]: So `TypeAttribs` and `ValueAttribs` here never referred to `TypedMorphirRuntimeDefs` (tho they share the same name as types defined there). As such, I think they should have been left unchanged.

File: morphir/runtime/src/org/finos/morphir/runtime/Runtime.scala [hunk 3]
```diff
@@ -1,14 +1,12 @@
 package org.finos.morphir
 package runtime
 
+import org.finos.morphir.ir.Type.UType
 import org.finos.morphir.naming._
 trait Runtime { self =>
   import Runtime._
 
-  type TypeAttribs
```
Comment (edwardpeters) [inline morphir/runtime/src/org/finos/morphir/runtime/Runtime.scala:8]: Similar to the above, these are unrelated to `TypedMorphirRuntimeDefs` and should not have been changed.

File: morphir/runtime/src/org/finos/morphir/runtime/quick/EvaluatorQuick.scala [hunk 4]
```diff
@@ -23,7 +23,7 @@ object EvaluatorQuick {
   type FloatType = Double
 
   private[runtime] def evalAction(
-      value: Value[TypeAttribs, ValueAttribs],
+      value: Value[scala.Unit, UType],
```
Comment (edwardpeters) [inline morphir/runtime/src/org/finos/morphir/runtime/quick/EvaluatorQuick.scala:26]: This case is more complicated: These did refer to `TypedMorphirRuntimeDefs`, but in a weird way where the `Value[_, _]` type was manually constructed. I think this should just be `value : TypedValue` - @DamianReeves , do you agree?

File: morphir/runtime/src/org/finos/morphir/runtime/quick/EvaluatorQuick.scala [hunk 5]
```diff
@@ -33,7 +33,7 @@ object EvaluatorQuick {
     }
 
   private[runtime] def eval(
-      value: Value[TypeAttribs, ValueAttribs],
+      value: Value[scala.Unit, UType],
```
Comment (edwardpeters) [inline morphir/runtime/src/org/finos/morphir/runtime/quick/EvaluatorQuick.scala:36]: See above re: `TypedValue`

File: morphir/runtime/src/org/finos/morphir/runtime/quick/Loop.scala [hunk 6]
```diff
@@ -271,9 +270,9 @@ private[morphir] case class Loop(globals: GlobalDefs) extends InvokeableEvaluato
   }
 
   def handleLetRecursion(
-      va: ValueAttribs,
-      definitions: Map[Name, Definition[TypeAttribs, ValueAttribs]],
-      inValue: RuntimeValue,
+      va: UType,
+      definitions: Map[Name, Definition[scala.Unit, UType]],
```
Comment (edwardpeters) [inline morphir/runtime/src/org/finos/morphir/runtime/quick/Loop.scala:274]: So this probably should have been `RuntimeDefinition` before, and should be `TypedDefinition` now.

BTW @Bouzomgi , this is what I briefly mentioned about none of these being enforced because none of the types are opaque - since the compiler can see that these all boil down to `scala.Unit` and `Type[scala.Unit]`, it wasn't enforcing anything about how we aliased or referenced them.

File: morphir/runtime/src/org/finos/morphir/runtime/quick/Loop.scala [hunk 7]
```diff
@@ -2,12 +2,11 @@ package org.finos.morphir.runtime.quick
 
 import org.finos.morphir.naming.*
```
Comment (edwardpeters) [inline morphir/runtime/src/org/finos/morphir/runtime/quick/Loop.scala:3]: @DamianReeves Other than the one `Definition[UType, Unit]` case I commented on, this file is exactly what I was expecting to see. Can you take a look and make sure this is also what you were expecting/agreed to?

File: morphir/src/org/finos/morphir/runtime/RTValue.scala [hunk 8]
```diff
@@ -2,22 +2,22 @@ package org.finos.morphir.runtime
...
 import org.finos.morphir.runtime.MorphirRuntimeError.{IllegalValue, FailedCoercion}
 import org.finos.morphir.runtime.internal.CallStackFrame
+import org.finos.morphir.ir.Type.UType
 
 import scala.collection.immutable.{List as ScalaList, Set as ScalaSet, Map as ScalaMap}
 import scala.collection.mutable
 
 // TODO Integrate errors into reporting format
-// Represents a Morphir-Evaluator result. Typed on TypedMorphirRuntimeDefs.TypeAttribs, TypedMorphirRuntimeDefs.ValueAttribs
+// Represents a Morphir-Evaluator result. Typed on TypedMorphirRuntimeDefs.scala.Unit, TypedMorphirRuntimeDefs.UType
```
Comment (edwardpeters) [inline morphir/src/org/finos/morphir/runtime/RTValue.scala:20]: This comment should probably say either TypedValue or Unit, UType

File: morphir/src/org/finos/morphir/runtime/RTValue.scala [hunk 9]
```diff
@@ -642,24 +642,24 @@ object RTValue {
...
 
-  case class LambdaFunction(body: RuntimeValue, pattern: Pattern[ValueAttribs], closingContext: CallStackFrame)
+  case class LambdaFunction(body: TypedValue, pattern: Pattern[UType], closingContext: CallStackFrame)
       extends Function
 
   case class DefinitionFunction(
-      body: RuntimeValue,
-      arguments: scala.List[(Name, ValueAttribs, Type[TypeAttribs])],
+      body: TypedValue,
+      arguments: scala.List[(Name, UType, Type[scala.Unit])],
```
Comment (edwardpeters) [inline morphir/src/org/finos/morphir/runtime/RTValue.scala:657]: This one is a bit weird. The two types specified here are the same - I had to do some digging, but that is in fact the case. It comes from Morphir's `Definition` type, which takes both value attributes and a type for each input, because `Definition`s are typed whether or not `VA` represents that type, but `VA` could potentially hold some other information. I copied that format over into my `DefinitionFunction`, despite the fact that it never uses those types (it trusts the type checker did its job and just counts the number of applied vs. expected arguments).

Cleaning that up is a broader discussion, tho, so I would just change this to `UType, UType` and leave it at that for now.

File: morphir/runtime/src/org/finos/morphir/runtime/Runtime.scala (hunk 3 above)
Comment (edwardpeters) [reply morphir/runtime/src/org/finos/morphir/runtime/Runtime.scala:8]: (I see a few places where that happened - anywhere where `TypeAttribs` and `ValueAttribs` are defined locally rather than imported from `TypedMorphirRuntimeDefs` we should leave be.)

File: morphir/runtime/src/org/finos/morphir/runtime/quick/EvaluatorQuick.scala (hunk 4 above)
Comment (DamianReeves) [reply morphir/runtime/src/org/finos/morphir/runtime/quick/EvaluatorQuick.scala:26]: Yeah I think that is fine for this one.

File: morphir/runtime/src/org/finos/morphir/runtime/quick/GatherReferences.scala [hunk 10]
```diff
@@ -13,10 +13,9 @@ import org.finos.morphir.runtime.SDKValue
 
 import scala.collection.immutable.Set
 import zio.Chunk
-import org.finos.morphir.runtime.TypedMorphirRuntimeDefs.{TypeAttribs, ValueAttribs}
 
 object GatherReferences {
-  type TypedValue = Value[TypeAttribs, ValueAttribs]
+  type TypedValue = Value[scala.Unit, UType]
```
Comment (Bouzomgi) [inline morphir/runtime/src/org/finos/morphir/runtime/quick/GatherReferences.scala:18]: @edwardpeters Does this line need to change? I am unsure of this one -- should I make it point to `Value.scala` 's typedValue instead? Like
import org.finos.morphir.ir.TypedValue as TV
`type TypedValue = TV`

Comment (edwardpeters) [reply morphir/runtime/src/org/finos/morphir/runtime/quick/GatherReferences.scala:18]: I would point it at `Value.scala`, and maybe skip the rename and just refer to everything as `TypedValue` directly.

File: morphir/runtime/src/org/finos/morphir/runtime/quick/EvaluatorQuick.scala [hunk 11]
```diff
@@ -9,12 +9,12 @@ import org.finos.morphir.ir.distribution.Distribution
 import org.finos.morphir.naming.*
 import org.finos.morphir.runtime.Extractors.*
 import org.finos.morphir.runtime.Extractors.Types.*
-import org.finos.morphir.runtime.TypedMorphirRuntimeDefs.{TypeAttribs, ValueAttribs}
 import org.finos.morphir.runtime.{Distributions, RTValue}
 import org.finos.morphir.runtime.MorphirRuntimeError.*
 import org.finos.morphir.runtime.environment.MorphirEnv
 import org.finos.morphir.runtime.exports.*
 import org.finos.morphir.runtime.services.sdk.*
+import org.finos.morphir.ir.Type.UType
```
Comment (Bouzomgi) [inline morphir/runtime/src/org/finos/morphir/runtime/quick/EvaluatorQuick.scala:17]: NOTE: Can remove

**Comment by Bouzomgi** [comment]
Hi @edwardpeters @DamianReeves , was a rough first go but I think this pass is much cleaner. Thank you for the feedback, hopefully my changes address your concerns.

File: morphir/runtime/src/org/finos/morphir/runtime/MorphirRuntime.scala [hunk 12]
```diff
@@ -9,34 +9,33 @@ import org.finos.morphir.runtime.environment.MorphirEnv
 import org.finos.morphir.runtime.exports.RTAction
 import org.finos.morphir.runtime.quick.QuickMorphirRuntime
 import org.finos.morphir.runtime.MorphirRuntimeError.*
+import org.finos.morphir.ir.Value.TypedValue
 
 trait MorphirRuntime {
-  type TypeAttribs
```
Comment (edwardpeters) [inline morphir/runtime/src/org/finos/morphir/runtime/MorphirRuntime.scala:14]: These still look wrong to me - the thing is, these never referred to `TypedMorphirRuntimeDefs` at all. I don't think anything in this file needed to change.

Comment (Bouzomgi) [reply morphir/runtime/src/org/finos/morphir/runtime/MorphirRuntime.scala:14]: Previously I attempted to keep these here, but in the file `TypedMorphirRuntimes`, it implements these signatures. It is expecting definitions for TypeAttribs + ValueAttribs, and it was pulling those definitions from the erradicated `TypedMorphirRuntimeDefs` . Since this file has been removed, there is no longer an implementation for TypeAttribs + ValueAttribs.

File: morphir/runtime/src/org/finos/morphir/runtime/TypedMorphirRuntime.scala [hunk 13]
```diff
@@ -12,23 +12,22 @@ import org.finos.morphir.datamodel.Util.*
 import org.finos.morphir.datamodel.*
 import org.finos.morphir.runtime.environment.MorphirEnv
 import org.finos.morphir.runtime.exports.*
+import org.finos.morphir.ir.Value.TypedValue
 
 trait TypedMorphirRuntime extends MorphirRuntime {
-  type TypeAttribs  = TypedMorphirRuntimeDefs.TypeAttribs
-  type ValueAttribs = TypedMorphirRuntimeDefs.ValueAttribs
```
Comment (edwardpeters) [inline morphir/runtime/src/org/finos/morphir/runtime/TypedMorphirRuntime.scala:18]: I think setting these to `Unit` and `UType` is reasonable.

File: morphir/runtime/src/org/finos/morphir/runtime/MorphirRuntime.scala (hunk 12 above)
Comment (edwardpeters) [reply morphir/runtime/src/org/finos/morphir/runtime/MorphirRuntime.scala:14]: Okay, I see. You still don't need to remove these definitions or change this file, but rather than removing the associated bindings in `TypedMorphirRuntimes`, you can probably just bind them to `Unit` and `UType`. (I added a comment on the relevant lines).

**Review by edwardpeters — APPROVED** [review]
LGTM

**Review by DamianReeves — APPROVED** [review]
LGTM
